pkg/registry/common/expire/nse_server_test.go

```diff
@@ -121,6 +125,35 @@ func TestExpireNSEServer_ShouldRemoveNSEAfterExpirationTime(t *testing.T) {
 	}, time.Second, time.Millisecond*100)
 }
 
+func TestExpireNSEServer_DataRace(t *testing.T) {
```

Comment:
This datarace fixes by more simple patch
```
diff --git a/pkg/registry/common/expire/nse_server.go b/pkg/registry/common/expire/nse_server.go
index a92044d..20b41bb 100644
--- a/pkg/registry/common/expire/nse_server.go
+++ b/pkg/registry/common/expire/nse_server.go
@@ -56,11 +56,10 @@ func (n *expireNSEServer) Register(ctx context.Context, nse *registry.NetworkSer
                defer cancel()
                _, _ = next.NetworkServiceEndpointRegistryServer(unregisterCtx).Unregister(unregisterCtx, unregisterNSE)
        })
-       if t, load := n.timers.LoadOrStore(resp.Name, timer); load {
+       if timer, ok := n.timers.LoadAndDelete(nse.Name); ok {
                timer.Stop()
-               t.Stop()
-               t.Reset(time.Until(expirationTime))
        }
+       n.timers.Store(resp.Name, timer)
 
        return resp, nil
 }
@@ -75,7 +74,7 @@ func (n *expireNSEServer) Unregister(ctx context.Context, nse *registry.NetworkS
                return nil, err
        }
 
-       if timer, ok := n.timers.Load(nse.Name); ok {
+       if timer, ok := n.timers.LoadAndDelete(nse.Name); ok {
                timer.Stop()
        }
 
```

Reply:
Yes, it fixes data race but there comes issue with possibly not stopped timers that can execute after the successful refresh:
```
1. timer fires unregister
2. refresh comes
3. refresh reaches memory
4. unregister reaches memory
```
in such case we will have no entry in memory right after the refresh happen.